url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/lists/PeList.java

```diff
@@ -0,0 +1,238 @@
+	 * 
+	 * @param id the Pe ID
+	 * @param peList the pe list
+	 * @return the MIPS rating if exists, otherwise returns -1
+	 * @pre id >= 0
+	 * @post $none
+	 */
+	public static <T extends Pe> Pe getById(List<Pe> peList, int id) {
+		for (Pe pe : peList) {
+			if (pe.getId() == id) {
+				return pe;
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Gets MIPS Rating for a specified Pe ID.
+	 * 
+	 * @param id the Pe ID
+	 * @param peList the pe list
+	 * @return the MIPS rating if exists, otherwise returns -1
+	 * @pre id >= 0
+	 * @post $none
+	 */
+	public static <T extends Pe> int getMips(List<Pe> peList, int id) {
+		Pe pe = getById(peList, id);
+		if (pe != null) {
+			return pe.getMips();
+		}
+		return -1;
+	}
+
+	/**
+	 * Gets total MIPS Rating for all PEs.
+	 * 
+	 * @param peList the pe list
+	 * @return the total MIPS Rating
+	 * @pre $none
+	 * @post $none
+	 */
+	public static <T extends Pe> int getTotalMips(List<Pe> peList) {
+		int totalMips = 0;
+		for (Pe pe : peList) {
+			totalMips += pe.getMips();
+		}
+		return totalMips;
+	}
+        
+        public static <T extends Pe> int getTotalMaxMips(List<Pe> peList) {
+		int totalMips = 0;
+		for (Pe pe : peList) {
+			totalMips += pe.getMaxMips();
+		}
+		return totalMips;
+	}
+
+	/**
+	 * Gets the max utilization among by all PEs.
+	 * 
+	 * @param peList the pe list
+	 * @return the utilization
+	 */
+	public static <T extends Pe> double getMaxUtilization(List<Pe> peList) {
+		double maxUtilization = 0;
+		for (Pe pe : peList) {
+			double utilization = pe.getPeProvisioner().getUtilization();
+			if (utilization > maxUtilization) {
+				maxUtilization = utilization;
+			}
+		}
+		return maxUtilization;
+	}
+
+	/**
+	 * Gets the max utilization among by all PEs allocated to the VM.
+	 * 
+	 * @param vm the vm
+	 * @param peList the pe list
+	 * @return the utilization
+	 */
+	public static <T extends Pe> double getMaxUtilizationAmongVmsPes(List<Pe> peList, Vm vm) {
+		double maxUtilization = 0;
+		for (Pe pe : peList) {
+			if (pe.getPeProvisioner().getAllocatedMipsForVm(vm) == null) {
+				continue;
+			}
+			double utilization = pe.getPeProvisioner().getUtilization();
+			if (utilization > maxUtilization) {
+				maxUtilization = utilization;
+			}
+		}
+		return maxUtilization;
+	}
+
+	/**
+	 * Gets a Pe ID which is FREE.
+	 * 
+	 * @param peList the pe list
+	 * @return a Pe ID if it is FREE, otherwise returns -1
+	 * @pre $none
+	 * @post $none
+	 */
+	public static <T extends Pe> Pe getFreePe(List<Pe> peList) {
+		for (Pe pe : peList) {
+			if (pe.getStatus() == Pe.FREE) {
+				return pe;
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Gets the number of <tt>FREE</tt> or non-busy Pe.
+	 * 
+	 * @param peList the pe list
+	 * @return number of Pe
+	 * @pre $none
+	 * @post $result >= 0
+	 */
+	public static <T extends Pe> int getNumberOfFreePes(List<Pe> peList) {
+		int cnt = 0;
+		for (Pe pe : peList) {
+			if (pe.getStatus() == Pe.FREE) {
+				cnt++;
+			}
+		}
+		return cnt;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 148:154, 186:192.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.